### ISSUE TITLE
Sync freezes when server snapshot split into parts

### ISSUE THREAD

**Issue by vunder** [Mimetis/Dotmim.Sync#333]
On server parts I have code to create snapshot
```csharp
        [HttpPost("CreateSnapshot")]
        public async Task<IActionResult> CreateSnapshotAsync()
        {
            var connectionString = configuration.GetConnectionString("DefaultConnection");
            var serverProvider = new SqlSyncProvider(connectionString);

            // Create a remote orchestrator
            var remoteOrchestrator = new RemoteOrchestrator(serverProvider, syncOptions, syncSetup);

            // Create a snapshot
            var sp = new SyncParameters();
            sp.Add("userId", User.FindFirst(ClaimTypes.NameIdentifier).Value);
            var bi = await remoteOrchestrator.CreateSnapshotAsync(sp);

            return Ok();
        }
```
Before starting sync I'm calling this method from client.
When sync is running in `Reinitialize` sync freezes while getting another table from batch (actually while deserializing another batch https://github.com/Mimetis/Dotmim.Sync/blob/302121209e6e9487f80fc104954e86d5ea2247bb/Projects/Dotmim.Sync.Core/Serialization/JsonConverter.cs#L33)
When `SyncOptions.BatchSize` is big enougth to have a single file for snapshot - no problems

**Comment by Mimetis** [comment]
You should not create a snapshot everytime.
The purpose of snapshot is to mitigate the server work.
Maybe once a week is enough.

I still don't understand why you have freezes.
And not sure it's related to Snapshots
Even if you have a snapshot or not, the web server will send a request that will be saved locally on the client (and then deserialized) 

So not sure the problem is coming from the snapshot

And everything is async by default, so not sure I understand what you mean by "freezing" ?

**Comment by vunder** [comment]
> The purpose of snapshot is to mitigate the server work.

I'v done in this way for testing purposes

> so not sure I understand what you mean by "freezing"?

Since `async` it is not freezing. But there is not more activity. Trying to debug step-by-step never passes deserialize of the second batch part.

**Comment by Mimetis** [comment]
can u share a sample repro ?

**Comment by vunder** [comment]
https://github.com/Mimetis/Dotmim.Sync/blob/302121209e6e9487f80fc104954e86d5ea2247bb/Projects/Dotmim.Sync.Core/Batch/BatchInfo.cs#L167-L168

**Comment by vunder** [comment]
> can u share a sample repro ?

Oh, I'll try

**Comment by Mimetis** [comment]
https://github.com/Mimetis/Dotmim.Sync/blob/302121209e6e9487f80fc104954e86d5ea2247bb/Projects/Dotmim.Sync.Core/Batch/BatchInfo.cs#L161-L168

Maybe I should try to use the compatiblity package, [Microsoft.Bcl.AsyncInterfaces](https://www.nuget.org/packages/Microsoft.Bcl.AsyncInterfaces)  as mentioned here : https://btburnett.com/csharp/2019/12/01/iasyncenumerable-is-your-friend.html#can-i-use-iasyncenumerable-when-targeting-net-core-2x

And see if it resolves the freezes you are mentioning.
I need a repro sample as well, so if you can provide something it will help a lot :)

**Comment by Mimetis** [comment]
@Vunder I've made a quick fix on the branch https://github.com/Mimetis/Dotmim.Sync/tree/IAsyncEnumerable_NetStandard20

Can you give it a try and see if you're still experimenting freezes ?

**Comment by vunder** [comment]
@Mimetis Here is a sample https://github.com/vunder/Dotmim.Sync/tree/batch-deadlock-sample/Samples/DeadlockSample
You will find server and two clients: Console and WPF. Console client works correct (I think in Intergation tests it would be impossible to catch this issue). But WPF gets into deadlock

**Comment by vunder** [comment]
@Mimetis Merged changes from `IAsyncEnumerable_NetStandard20` branch. Issue wixed. Also, you can update C# version within project to `8.0` and use following syntax
```csharp
                await foreach (var syncTable in message.Changes.GetTableAsync(schemaTable.TableName, schemaTable.SchemaName, this.Orchestrator))
```

**Comment by Mimetis** [comment]
I've upgraded to `C# 8.0` but we are still on `.Net Standard 2.0`
So far we can't use all the new cool things in `C# 8.0`  like Async Iterators, so we have to use iterator, like here :
https://btburnett.com/csharp/2019/12/01/iasyncenumerable-is-your-friend.html#can-i-use-iasyncenumerable-when-targeting-net-core-2x

Anyway, good to know that the issue is fixed :)

Will make a merge asap

**Comment by Mimetis** [comment]
@vunder I've made some twicks in the [MariaDB branch](https://github.com/Mimetis/Dotmim.Sync/tree/MariaDB) to ensure we have the less amount of freezes as possible.
Let me know if you have perf / ui improvements from this commit

I still have a bug on this branch for the MariaDB part, but once it's resolved I will merge into master

**Comment by Mimetis** [comment]
`MariaDB` branch merged into `master` :)
